ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -53,25 +53,58 @@ let sync_required ~__context ~host =
     end
   | _ -> raise Api_errors.(Server_error (internal_error, ["Cannot have more than one Cluster object per pool currently"]))
 
+let call_api_function_with_alert ~__context ~msg ~cls ~obj_uuid ~body ~(api_func : (Rpc.call -> Rpc.response) -> API.ref_session -> 'a) =
+  Helpers.call_api_functions ~__context
+    (fun rpc session_id ->
+       try
+         api_func rpc session_id |> ignore
+       with err ->
+         error "%s" ExnHelper.(string_of_exn err);
+         Xapi_alert.add ~msg ~cls ~obj_uuid ~body;
+         raise err
+    )
+
 let create_as_necessary ~__context ~host =
   match sync_required ~__context ~host with
   | Some cluster_ref -> (* assume pool autojoin set *)
     let network = get_network_internal ~__context ~self:cluster_ref in
     let (pifref,pifrec) = Xapi_clustering.pif_of_host ~__context network host in
     fix_pif_prerequisites ~__context (pifref,pifrec);
-    Helpers.call_api_functions ~__context (fun rpc session_id ->
-        Client.Client.Cluster_host.create rpc session_id cluster_ref host pifref) |> ignore
+    let api_func =
+      (fun rpc session_id ->
+         Client.Client.Cluster_host.create rpc session_id cluster_ref host pifref
+      ) in
+    let body =
+      (Printf.sprintf
+         "Unable to create cluster host on %s."
+         (Db.Host.get_name_label ~__context ~self:host)
+      ) in
+    let obj_uuid = Db.Host.get_uuid ~__context ~self:host in
+    call_api_function_with_alert ~__context
+      ~msg:Api_messages.cluster_host_creation_failed
+      ~cls:`Host ~obj_uuid ~body ~api_func
   | None -> ()
 
 let resync_host ~__context ~host =
   create_as_necessary ~__context ~host;
   match (find_cluster_host ~__context ~host) with
-    | None              -> () (* no clusters exist *)
-    | Some cluster_host ->    (* cluster_host and cluster exist *)
-      (* Cluster_host.enable unconditionally invokes the low-level enable operations and is idempotent. *)
-      if Db.Cluster_host.get_enabled ~__context ~self:cluster_host
-      then Helpers.call_api_functions ~__context
-        (fun rpc session_id -> Client.Client.Cluster_host.enable ~rpc ~session_id ~self:cluster_host)
+  | None              -> () (* no clusters exist *)
+  | Some cluster_host ->    (* cluster_host and cluster exist *)
+    (* Cluster_host.enable unconditionally invokes the low-level enable operations and is idempotent. *)
+    if Db.Cluster_host.get_enabled ~__context ~self:cluster_host then
+      let api_func =
+        (fun rpc session_id ->
+           Client.Client.Cluster_host.enable ~rpc ~session_id ~self:cluster_host
+        ) in
+      let body =
+        (Printf.sprintf
+           "Unable to create cluster host on %s"
+           (Db.Host.get_name_label ~__context ~self:host)
+        ) in
+      let obj_uuid = Db.Host.get_uuid ~__context ~self:host in
+      call_api_function_with_alert ~__context
+        ~msg:Api_messages.cluster_host_creation_failed
+        ~cls:`Host ~obj_uuid ~body ~api_func
```

Comment:
`api_func` could probably just be declared directly here, like it was for `Helpers.call_api_functions`. i.e. you would replace the Helpers.call_api_functions with `call_api_function_with_alert`.